src/schemas/generateSchema.ts

```diff
@@ -0,0 +1,112 @@
+#!usr/bin/env node
+require('dotenv').config();
+import 'module-alias/register';
+
+import fs from 'fs';
+import path from 'path';
+import { ensureEnvVariable } from '../utils';
+
+export const generateSchema = async () => {
+  const name = ensureEnvVariable('CHAIN');
+  const indexerFile = path.join(__dirname, '../chain', name, '/main.js');
+
+  const { indexer } = await import(indexerFile);
+
+  const eventKeys: string[] = [...indexer.mapper.events.keys()];
+  const callKeys: string[] = [...indexer.mapper.calls.keys()];
+
+  const combinedKeysSorted: string[] = [...eventKeys, ...callKeys].sort();
+  const schemaPath = path.resolve(__dirname, '../../schema.graphql');
+
+  buildSchema(combinedKeysSorted, schemaPath);
+};
+
+// NOTE: This is done to determine if the file is being run directly or being imported as a module
+if (require.main === module) {
+  generateSchema().catch((error) => {
+    console.error(error);
+    process.exit(1);
+  });
+}
+
+const buildSchema = (chainPalletKeys: string[], schemaPath: string) => {
+  const commentsInSchema = `# This is an auto-generated file used by the graphiQL explorer to define the tables and relationships between them.\n# This file will be auto-populated based on the pallets used in src/chain/{chainName}/main.ts\n# See generateSchema.ts for more information on how this file is generated.\n\n`;
+  const queryLogs = fs.readFileSync(path.join(__dirname, 'queryLogs.graphql'), 'utf8');
+
+  // Clear the file before starting the loop and add essential files and comments
+  fs.writeFileSync(schemaPath, '');
+  fs.appendFileSync(schemaPath, commentsInSchema);
+  fs.appendFileSync(schemaPath, queryLogs);
+  fs.appendFileSync(schemaPath, '\n');
+
+  const appendedSchemaParts = new Set<string>();
+  let accountSchema = `type Account @entity {\nid: ID!\npublicKey: ID! @index\n`;
```

Comment:
it's probably better to have this defined as an array to which you call `.join()` at the end as it does tend to outperform concatenation with +. Needless to say this goes for all such occurrences if there are more similar files